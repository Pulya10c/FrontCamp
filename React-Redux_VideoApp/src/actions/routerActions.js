import { history } from 'src/configureStore';

export const goTo = (routePath, searchStr = '') => {
  history.push({
    pathname: routePath,
    search: searchStr
  });
};
