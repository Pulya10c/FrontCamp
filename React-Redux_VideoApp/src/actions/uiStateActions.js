import { filterConfigSelector } from 'selectors/videoListSelectors';

const MODULE_NAME = 'UI';

export const START_FETCHING = `${MODULE_NAME}/START_FETCHING`;
export const FINISH_FETCHING = `${MODULE_NAME}/FINISH_FETCHING`;
export const SET_ERROR = `${MODULE_NAME}/SET_ERROR`;

export const startFetching = () => {
  return {
    type: START_FETCHING
  };
};

export const finishFetching = () => {
  return {
    type: FINISH_FETCHING
  };
};

export const setError = error => {
  return {
    type: SET_ERROR,
    payload: error
  };
};
