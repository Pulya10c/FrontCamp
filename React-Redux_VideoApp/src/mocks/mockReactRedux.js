const createMockMapStateToProps = mapStateToProps => {
  if (!mapStateToProps) {
    return mapStateToProps;
  }

  return (state, ownProps) => {
    return {
      props: mapStateToProps(state, ownProps)
    };
  };
};

const getDispatchFunctionFormShorthandObject = actionsObj => dispatch => {
  return Object.keys(actionsObj).reduce((props, actionKey) => {
    return {
      ...props,
      [actionKey]: function() {
        return dispatch(actionsObj[actionKey].apply(this, arguments));
      }
    };
  }, {});
};

const createMockMapDispatchToProps = mapDispatchToProps => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps;
  }

  const mappingFunction =
    typeof mapDispatchToProps === 'function'
      ? mapDispatchToProps
      : getDispatchFunctionFormShorthandObject(mapDispatchToProps);

  return (dispatch, ownProps) => {
    dispatch = dispatch || jest.fn(action => action);

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    };
  };
};

const createMockMergeProps = mergeProps => {
  if (!mergeProps) {
    return mergeProps;
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: jest.fn(action => action) };

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    };
  };
};

export default {
  connect: (
    mapStateToProps,
    mapDispatchToProps,
    mergeProps
  ) => ConnectedComponent => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    ConnectedComponent
  })
};
