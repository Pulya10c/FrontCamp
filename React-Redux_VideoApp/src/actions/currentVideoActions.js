import QueryStringGenerator from 'services/QueryStringGenerator';
import RequestService from 'services/RequestService';
import { updateFilterConfigBy, fetchMovies } from 'actions/videoListActions';
import { API, GENRES, TITLE, PATHS } from 'src/constants';
import { currentVideoGenresSelector } from 'selectors/currentVideoSelectors';
import * as uiActions from 'actions/uiStateActions';
import { goTo } from 'actions/routerActions';

const MODULE_NAME = 'CURRENT_VIDEO';

export const SAVE_FILM = `${MODULE_NAME}/SAVE_FILM`;
export const SET_SELECTED_FILM_ID = `${MODULE_NAME}/SET_SELECTED_FILM_ID`;

// export for test
export const _saveFilm = film => {
  return {
    type: SAVE_FILM,
    payload: film
  };
};

export const fetchCurrentVideoAndSimilarFilms = id => async (
  dispatch,
  getState
) => {
  try {
    const queryString = QueryStringGenerator.getUrlForRequest(id);
    dispatch(uiActions.startFetching());
    const film = await RequestService.getResponse(queryString);
    dispatch(uiActions.finishFetching());
    dispatch(_saveFilm(film));
    const state = getState();
    const genres = currentVideoGenresSelector(state);
    dispatch(updateFilterConfigBy[API.SEARCH](genres.join(' ')));
    dispatch(updateFilterConfigBy[API.SEARCH_BY](GENRES));
    dispatch(fetchMovies());
  } catch (error) {
    console.error(error);
    dispatch(uiActions.setError(error));
  }
};

export const resetSearchAngRedirect = () => dispatch => {
  dispatch(updateFilterConfigBy[API.SEARCH](''));
  dispatch(updateFilterConfigBy[API.SEARCH_BY](TITLE.code));
  goTo(PATHS.VIDEO_LIST_PAGE);
};
