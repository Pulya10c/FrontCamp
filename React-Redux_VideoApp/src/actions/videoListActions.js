import has from 'lodash/has';

import { filterConfigSelector } from 'selectors/videoListSelectors';
import QueryStringGenerator from 'services/QueryStringGenerator';
import RequestService from 'services/RequestService';
import * as uiActions from 'actions/uiStateActions';
import { goTo } from 'actions/routerActions';
import { routerSearchSelector } from 'selectors/routerSelectors';
import { PATHS, API, GENRES } from 'src/constants';

const MODULE_NAME = 'VIDEO_LIST';

export const SET_SEARCH_VALUE = `${MODULE_NAME}/SET_SEARCH_VALUE`;
export const SET_SEARCH_BY_VALUE = `${MODULE_NAME}/SET_SEARCH_BY_VALUE`;
export const SET_SORT_ORDER_VALUE = `${MODULE_NAME}/SET_SORT_ORDER_VALUE`;
export const SET_SORT_BY_VALUE = `${MODULE_NAME}/SET_SORT_BY_VALUE`;
export const SET_OFFSET_VALUE = `${MODULE_NAME}/SET_OFFSET_VALUE`;
export const SET_LIMIT_VALUE = `${MODULE_NAME}/SET_LIMIT_VALUE`;

export const SAVE_MOVIES = `${MODULE_NAME}/SAVE_MOVIES`;
export const UPDATE_FILTER_FROM_SEARCH_PARAMS = `${MODULE_NAME}/UPDATE_FILTER_FROM_SEARCH_PARAMS`;

export const updateFilterConfigBy = {
  [API.SEARCH_BY]: value => ({
    type: SET_SEARCH_BY_VALUE,
    payload: value
  }),
  [API.SEARCH]: value => ({
    type: SET_SEARCH_VALUE,
    payload: value
  }),
  [API.SORT_ORDER]: value => ({
    type: SET_SORT_ORDER_VALUE,
    payload: value
  }),
  [API.SORT_BY]: value => ({
    type: SET_SORT_BY_VALUE,
    payload: value
  }),
  [API.OFFSET]: value => ({
    type: SET_OFFSET_VALUE,
    payload: value
  }),
  [API.LIMIT]: value => ({
    type: SET_LIMIT_VALUE,
    payload: value
  })
};

// export for tests
export const _saveMovies = movies => {
  return {
    type: SAVE_MOVIES,
    payload: movies
  };
};

const _updateFilterFromSearchParams = () => ({
  type: UPDATE_FILTER_FROM_SEARCH_PARAMS
});

// export for tests
export const _updateFilterConfigInOrderToSearchParams = (
  searchStr,
  filterConfig
) => dispatch => {
  const searchParams = new URLSearchParams(searchStr);
  for (const [key, value] of searchParams) {
    if (has(filterConfig, key) && filterConfig[key] != value) {
      dispatch(_updateFilterFromSearchParams());
      dispatch(updateFilterConfigBy[key](value));
    }
  }
};

export const fetchMovies = () => async (dispatch, getState) => {
  try {
    const state = getState();
    const searchStr = routerSearchSelector(state);
    const filterConfig = filterConfigSelector(state);
    dispatch(_updateFilterConfigInOrderToSearchParams(searchStr, filterConfig));
    const queryString = QueryStringGenerator.getUrlForRequest(searchStr);
    dispatch(uiActions.startFetching());
    const movies = await RequestService.getResponse(queryString);
    dispatch(uiActions.finishFetching());
    dispatch(_saveMovies(movies));
  } catch (error) {
    console.error(error);
    dispatch(uiActions.setError(error));
  }
};

export const createSearchAndRedirect = () => (dispatch, getState) => {
  const state = getState();
  const filterConfig = filterConfigSelector(state);
  const searchStr = QueryStringGenerator.getSearchStr(filterConfig);
  goTo(PATHS.VIDEO_LIST_PAGE, searchStr);
};

export const sortByValueAndFetchMovies = value => dispatch => {
  dispatch(updateFilterConfigBy[API.SORT_BY](value));
  dispatch(createSearchAndRedirect());
};
