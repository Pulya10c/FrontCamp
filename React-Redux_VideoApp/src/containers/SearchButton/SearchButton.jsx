import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import Button from "components/Button";
import { createSearchAndRedirect } from "actions/videoListActions";

const mapDispatchToProps = dispatch => ({
  createSearchAndRedirect: () => dispatch(createSearchAndRedirect())
});

const SearchButton = ({ createSearchAndRedirect }) => {
  return <Button btnTitle={"Search"} executeAction={createSearchAndRedirect} />;
};

SearchButton.propTypes = {
  createSearchAndRedirect: PropTypes.func
};

export default connect(null, mapDispatchToProps)(SearchButton);
