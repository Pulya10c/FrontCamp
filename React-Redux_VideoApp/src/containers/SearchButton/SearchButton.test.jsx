import React from "react";
import { shallow } from "enzyme";
import mockReactRedux from "mocks/mockReactRedux";
import { createSearchAndRedirect } from "actions/videoListActions";

import SearchButton from "./SearchButton";

jest.mock("react-redux", () => mockReactRedux);

const mockCreateSearchAndRedirect = "mockCreateSearchAndRedirect";

jest.mock("actions/videoListActions", () => ({
  createSearchAndRedirect: jest.fn(() => mockCreateSearchAndRedirect)
}));

describe("Component SearchButton", () => {
  const { mapDispatchToProps, ConnectedComponent } = SearchButton;

  it("should render correctly", () => {
    expect(shallow(<ConnectedComponent />)).toMatchSnapshot();
  });

  it("should pass correct props to component from mapDispatchToProps", () => {
    const { props, dispatch } = mapDispatchToProps();
    props.createSearchAndRedirect();
    expect(createSearchAndRedirect).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(mockCreateSearchAndRedirect);
  });
});
