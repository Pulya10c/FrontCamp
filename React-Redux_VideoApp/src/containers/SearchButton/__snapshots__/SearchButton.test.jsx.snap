// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SearchButton should render correctly 1`] = `
<ButtonComponent
  btnTitle="Search"
/>
`;
